crates/evm-tracing-events/src/marshalled_opcode.rs

```diff
@@ -0,0 +1,252 @@
+//! Marshalled opcode definition and implementations.
+
+extern crate alloc;
+
+use codec::{Decode, Encode};
+use smallvec::SmallVec;
+use sp_core::sp_std::vec::Vec;
+
+/// Marshalled opcode.
+///
+/// 8 cause the longest is 13 and the epmeric estimate of max length
+/// for the most popular ones in 7-ish.
+#[derive(Default, Debug, Clone, PartialEq, Eq)]
+pub struct MarshalledOpcode(SmallVec<[u8; 8]>);
+
+impl From<&evm::Opcode> for MarshalledOpcode {
+    fn from(opcode: &evm::Opcode) -> Self {
+        MarshalledOpcode(SmallVec::from_vec(opcode_to_string(opcode)))
+    }
+}
+
+impl From<&'static str> for MarshalledOpcode {
+    fn from(value: &'static str) -> Self {
+        MarshalledOpcode(SmallVec::from_slice(value.as_bytes()))
+    }
+}
+
+impl Encode for MarshalledOpcode {
+    fn encode(&self) -> Vec<u8> {
+        self.0.clone().to_vec().encode()
+    }
+}
+
+impl Decode for MarshalledOpcode {
+    fn decode<I: codec::Input>(input: &mut I) -> Result<Self, codec::Error> {
+        let bytes = Vec::decode(input)?;
+        Ok(MarshalledOpcode(SmallVec::from_vec(bytes)))
+    }
+}
```

Comment:
Needs more optimized encode/decode

Reply:
Can you suggest something as I spent a lot of time trying different ways and met some issues with the encoding and decoding result data.

Then I decided to walk through `codec` implementation and this approach comes based on implementation for `String`.

Reply:
fixed type - slice `[u8]`

Reply:
Check whether `Vec<u8>` is used under the hood anywhere there - afaik borrow for `[u8]` is `Vec<u8>`; then it won't work